### PR TITLE
More Intuitive CLI for Static Transform Publisher

### LINKED ISSUE #295: static_transform_publisher more intuitive command-line arguments
## Bug report

The order of command-line arguments in the `static_transform_publisher` program is hard to remember and follow.  This led directly to the problem described in https://github.com/ros2/geometry2/issues/292.  It would be better if the command-line arguments were much more specific so there is no ambiguity.  We discussed this in https://github.com/ros2/geometry2/issues/292#issuecomment-668054758 and https://github.com/ros2/geometry2/issues/292#issuecomment-668732629 , and the consensus was to add the following command-line arguments:

Group 1:
* `--roll` - specify the roll in radians
* `--pitch` - specify the pitch in radians
* `--yaw` - specify the yaw in radians

Group 2:
* `--roll_deg` - specify the roll in degrees
* `--pitch_deg` - specify the pitch in degrees
* `--yaw_deg` - specify the yaw in degrees

Group 3:
* `--qx` - Specify the quaternion X
* `--qy` - Specify the quaternion Y
* `--qz` - Specify the quaternion Z
* `--qw` - Specify the quaternion W

Note that each of the groups is mutually exclusive; if the user specifies `--roll`, they cannot then specify `--roll_deg` or `--qx` (while we technically could support mixed `--roll` and `--pitch_deg`, we'll just specify that they are exclusive for simplicity).

In order to "tick-tock" this, we'll also continue to support the old command-line options up to and including Galactic, though with a deprecation warning.  After Galactic, we'll remove the old command-line options.  The old command-line options should also be mututally exclusive with any of the 3 groups above.

@allenh1 has graciously offered to take on implementing this.  @Stapelzeiger FYI.

### PR DESCRIPTION
**Pull request by allenh1**
Opening for visibility. Few TODO's left in line that I'll get to in a bit, but this fixes #295.

Signed-off-by: Hunter L. Allen <hunter.allen@ghostrobotics.io>

### CODE REVIEW

**Comment by allenh1** [comment]
RPY now supported, but need to add `--roll-deg` and the like still.

File: tf2_ros/src/static_transform_broadcaster_program.cpp [hunk 1]
```diff
@@ -35,68 +35,172 @@
...
+#include <utility>
 #include <vector>
 
+std::unordered_map<std::string, std::string> _make_arg_map(std::vector<std::string> && args)
+{
+  std::unordered_map<std::string, std::string> ret;
+  /* collect from [exe] --option1 value --option2 value ... --optionN value */
+  for (size_t x = 1; x < args.size(); x += 2) {
+    ret.emplace(std::move(args[x]), std::move(args[x + 1]));
+  }
```
Comment (clalancette) [inline tf2_ros/src/static_transform_broadcaster_program.cpp]: What happens here if the user passes an incorrect command line, like:

```
--x --y 0
```

File: tf2_ros/src/static_transform_broadcaster_program.cpp [hunk 2]
```diff
@@ -35,68 +35,172 @@
...
-    printf("OR \n");
-    printf("Usage: static_transform_publisher x y z yaw pitch roll frame_id child_frame_id \n");
-    RCUTILS_LOG_ERROR(
-      "static_transform_publisher exited due to not having the right number of arguments");
-    return 2;
+  tf2::Quaternion rotation;
+  tf2::Vector3 translation;
+  try {
+    rotation = _get_rotation(arg_map);
+    translation = _get_translation(arg_map);
```
Comment (clalancette) [inline tf2_ros/src/static_transform_broadcaster_program.cpp]: What happens if the user supplies arguments that we don't understand, like `--foo`?

File: tf2_ros/src/static_transform_broadcaster_program.cpp [hunk 3]
```diff
@@ -35,68 +35,172 @@
...
-    frame_id = args[7];
-    child_id = args[8];
-  } else {
-    // quaternion supplied directly
-    rx = std::stod(args[4]);
-    ry = std::stod(args[5]);
-    rz = std::stod(args[6]);
-    rw = std::stod(args[7]);
-    frame_id = args[8];
-    child_id = args[9];
```
Comment (clalancette) [inline tf2_ros/src/static_transform_broadcaster_program.cpp:79]: I think we still need to keep this old command-line parsing around for backwards-compatibility reasons.  My suggestion is that we first look for the new arguments; as we find them in the map, we remove them from the map.  Once we are done removing the new options, we look to see if there are 9 or 10 arguments left.  If there are, we print a warning that this usage is deprecated, but continue on anyway.  If both the new arguments and the old arguments are provided, we error and exit.  Does that seem like a reasonable plan?

**Review by clalancette — CHANGES_REQUESTED** [review]

File: tf2_ros/src/static_transform_broadcaster_program.cpp (hunk 1 above)
Comment (allenh1) [reply tf2_ros/src/static_transform_broadcaster_program.cpp]: Ah, that's a good catch. Would likely crash.

File: tf2_ros/src/static_transform_broadcaster_program.cpp (hunk 2 above)
Comment (allenh1) [reply tf2_ros/src/static_transform_broadcaster_program.cpp]: I think this is actually immune to that. It should set to the identity transform.

That said, it should probably error, since that's not a valid argument.

File: tf2_ros/src/static_transform_broadcaster_program.cpp (hunk 3 above)
Comment (allenh1) [reply tf2_ros/src/static_transform_broadcaster_program.cpp:79]: Yeah, that's absolutely a reasonable plan. Will do.

**Comment by tfoote** [comment]
For simplicity I'd suggest that instead of adding a bunch of extra logic to this command line parsing that we consider making a new executable with a new name which can take the new argument styles. This will ensure that we maintain the existing CLI and we can just reuse the existing implementation with a new command line parsing.

One simple alternative that gives a lot of power would be to actually make this into a python tool that has all the bells and whistles turned on, that then executes the less user friendly one under the hood. This would make the command line parsing much less error prone. And users who don't want python on their system can still use the less user friendly C++ version.

**Comment by clalancette** [comment]
> For simplicity I'd suggest that instead of adding a bunch of extra logic to this command line parsing that we consider making a new executable with a new name which can take the new argument styles. This will ensure that we maintain the existing CLI and we can just reuse the existing implementation with a new command line parsing.

The thing is, we still need most of the logic to parse the command-line anyway.  Adding a second executable only removes the backwards compatibility logic, and I don't expect that to be a whole lot of code anyway.  To me, the benefit of keeping the current executable name, and being able to (eventually) remove the old arguments outweighs the downside of carrying a bit of extra logic around.

I guess the command-line parsing becomes much easier if we switch the whole thing to Python.  But the thing is that we've gone a long way towards separating C++ and Python packages in geometry2, and I'd like to not go back on that.  So we'd have to move the new executable to `tf2_ros_py`, and then it is harder for users to migrate and discover, etc.  I just don't think that parsing the command-line arguments in C++ is going to be *that* hard, and so I don't think it warrants a big hammer here.

TLDR: my opinion is that we should continue down the path of parsing the command-line arguments in the current executable, and add (just a small bit more) logic to deal with the backwards-compatible arguments.

**Comment by tfoote** [comment]
Ok, If we were to switch it to be python based it would likely move to tf2_tools or the like to avoid the dependency.

**Comment by clalancette** [comment]
> Ok, If we were to switch it to be python based it would likely move to tf2_tools or the like to avoid the dependency.

Yes, that's a good point.

It seems to me that we have 3 options here:
1. Rewrite this utility to be in Python.  This has some obvious code maintenance benefits.  The downside is that it is disruptive to users on two axes: they need to change the package that the utility comes from, and they need to change the command-line arguments to the new syntax.
2. Create a new executable in C++ that only takes the new arguments.  This gets rid of some back-compat logic, but still requires that we write the command-line parsing ourselves.  This is disruptive to users on two axes: they need to change the executable they are running, and they need to change the command-line arguments to the new syntax.
3. Put all of the command-line parsing and back-compat right into the existing executable.  This is disruptive to users on one axis: they need to change the command-line arguments to the new syntax.

Option 3 is the least disruptive to users; option 1 is easier long-term maintenance (2 seems to be the worst of both worlds, so I won't consider it further).  Hard to choose :).

@allenh1 You are the one actually working on this, so any thoughts from you on this?

**Comment by allenh1** [comment]
> @allenh1 You are the one actually working on this, so any thoughts from you on this?

Hm. I think @tfoote makes a really good point here. I am not opposed to option 1 here at all.

I'm really not a fan of option 2. I can see that being a real maintenance headache in the future where we have the component version, the standalone executable, and the other standalone executable for the same thing. That's probably overkill, imo.

As for option 3, it's definitely a lot of code for setting arguments, but I'm much faster with my C++ than my Python to be honest. And I feel like after printing the deprecation warning for the old style in Galactic, it can be removed completely in H-turtle.

So my preference is definitely 3, but probably just because it's all C++. I am fine with 1 if that's what we want instead.

**Comment by clalancette** [comment]
@allenh1 I'll leave the final decision to you.  I'm slightly more in favor of 3 than 1, just because of the relative ease of transition.  Tully seems to be more in favor of 1 than 3, because of the maintenance.  You are doing the work, so you can do what you feel is best and be the tie-breaker :).

If you do decide to go with 1, remember that in addition to the new Python executable we'll want a small update to the existing static_transform_broadcaster_program to say `This utility is deprecated, use <new_executable> now`.

**Comment by allenh1** [comment]
> If you do decide to go with 3, remember that in addition to the new Python executable

Do you mean 1? Just to be clear...

**Comment by clalancette** [comment]
> Do you mean 1? Just to be clear...

Yes!  Sorry.  Will update the comment.

**Comment by clalancette** [comment]
@allenh1 I spent a bit of time here and reimplemented this to take care of all of the issues earlier identified, as well as adding support for both new-style and old-style arguments (the latter print a warning).  I think this now fulfills all of the requirements we were looking for; can you take a look at the code and let me know what you think about it?

**Comment by allenh1** [comment]
>  I think this now fulfills all of the requirements we were looking for; can you take a look at the code and let me know what you think about it?

@clalancette looks good to me. Thanks for finishing this up for me.

**Comment by clalancette** [comment]
@ahcorde When you get a chance, can you do a review of this?  Thanks.

**Review by clalancette — APPROVED** [review]
This looks good to me, but since I rewrote substantial portions of it I'll wait on @ahcorde before merging it.  I'll run CI on it in the meantime.

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15358/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10068/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13044/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15562/)

**Comment by clalancette** [comment]
Fixed some warnings on Windows and rebased.  Here's another try at CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15366/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10073/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13050/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15569/)

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15391/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10098/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13078/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15604/)

File: tf2_ros/src/static_transform_broadcaster_program.cpp [hunk 4]
```diff
@@ -28,76 +28,383 @@
...
+
+  bool saw_quat_flag = false;
+  bool saw_rpy_flag = false;
+  double roll = 0.0;
+  double pitch = 0.0;
+  double yaw = 0.0;
+
+  bool saw_trans_flag = false;
+
+  bool saw_frame_flag = false;
```
Comment (ahcorde) [inline tf2_ros/src/static_transform_broadcaster_program.cpp]:
```suggestion
  bool saw_frame_flag = false;
  bool saw_quat_flag = false;
  bool saw_rpy_flag = false;
  bool saw_trans_flag = false;
  double roll = 0.0;
  double pitch = 0.0;
  double yaw = 0.0;

```

**Review by ahcorde — APPROVED** [review]
just a mirnor style comment, otherwise LGTM

File: tf2_ros/src/static_transform_broadcaster_program.cpp (hunk 4 above)
Comment (clalancette) [reply tf2_ros/src/static_transform_broadcaster_program.cpp]: Good call, I've fixed this in e2d83c2

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15405/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10109/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13091/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15621/)

**Comment by clalancette** [comment]
One more CI to take the test warnings into account:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15408/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10112/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13094/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15626/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/15627/)

**Comment by clalancette** [comment]
CI is green, approved, so merging.  Thanks for the initial code @allenh1 , and thanks for the review @ahcorde .
